rust/parquet/src/record/mod.rs

```diff
@@ -16,9 +16,209 @@
+/// | [`Enum`](self::types::Enum) | byte_array | [enum](https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#enum) |
+/// | [`List<T>`](self::types::List) | group | [list](https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#lists) |
+/// | [`Map<K,V>`](self::types::Map) | group | [map](https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#maps) |
+/// | [`Group`](self::types::Group) | group | none |
+/// | `(T, U, …)` | group | none |
+/// | `Option<T>` | – | – |
+/// | [`Value`](self::types::Value) | * | * |
+///
+/// `Option<T>` corresponds to a field marked as "optional".
+///
+/// [`List<T>`](self::types::List) corresponds to either [annotated List logical types](https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#lists), or unannotated fields marked as "repeated".
+///
+/// [`Value`](self::types::Value) corresponds to any valid Parquet type, and is useful
+/// when the type is not known at compile time.
+///
+/// "byte_array" is interchangeable with "fixed_len_byte_array" for the purposes of the
+/// above correspondance.
+///
+/// The implementation for tuples is only for those up to length 32. The implementation
+/// for arrays is only for common array lengths. See [`Record`] for more details.
+///
+/// ## `#[derive(Record)]`
+///
+/// The easiest way to implement `Record` on a new type is using `#[derive(Record)]`:
+///
+/// ```
+/// use parquet::record::{types::Timestamp, Record};
+///
+/// #[derive(Record, Debug)]
+/// struct MyRow {
+///     id: u64,
+///     time: Timestamp,
+///     event: String,
+/// }
+/// ```
+///
+/// If the Rust field name and the Parquet field name differ, say if the latter is not an idiomatic or valid identifier in Rust, then an automatic rename can be made like so:
+///
+/// ```
+/// # use parquet::record::{types::Timestamp, Record};
+/// #[derive(Record, Debug)]
+/// struct MyRow {
+///     #[parquet(rename = "ID")]
+///     id: u64,
+///     time: Timestamp,
+///     event: String,
+/// }
+/// ```
+pub trait Record: Sized {
+    type Schema: Schema;
+    type Reader: Reader<Item = Self>;
+
+    /// Parse a [`Type`] into `Self::Schema`, using `repetition` instead of
+    /// `Type::get_basic_info().repetition()`. A `repetition` of `None` denotes a root
+    /// schema.
+    fn parse(
+        schema: &Type,
+        repetition: Option<Repetition>,
+    ) -> Result<(String, Self::Schema)>;
+
+    /// Builds tree of [`Reader`]s for the specified [`Schema`] recursively.
+    fn reader(
+        schema: &Self::Schema,
+        path: &mut Vec<String>,
+        def_level: i16,
+        rep_level: i16,
+        paths: &mut HashMap<ColumnPath, ColumnReader>,
+        batch_size: usize,
+    ) -> Self::Reader;
+}
+
+/// This trait is implemented by Schemas so that they can be printed as Parquet schema
+/// strings.
+pub trait Schema: Debug {
```

Comment:
Why we need this? only for displaying purpose? can we reuse `Type`?

Reply:
If the user uses the wrong type with `file_reader.get_row_iter<T>()` then ideally the error message includes the actual schema of the file, as well as the schema dictated by their type. This lets the user see where they differ and thus what's wrong.

The `Schema` trait exists so that a schema can be printed from just their type. This wasn't possible using `Type` as that requires a full actual schema, which cannot be attained from just the type.

Reply:
OK. It's a little confusing that the `fmt` is a public interface for the `Schema` type though - to someone who is new to this it may be hard to use (e.g., what value should I pass to the `r` and `name` parameters).